### PR TITLE
Use panicOnBuild instead of process.exit

### PR DESCRIPTION
**Pull request by andrewnester**
Observing the code I found out TODO item to use `reporter.panicOnBuild` instead of direct call to `process.exit`.

This PR actually implements this change.

### CODE REVIEW

**Comment by pieh** [comment]
Maybe we should call `panicOnBuild` directly here https://github.com/gatsbyjs/gatsby/blob/5b24e1acbaff9d1e42690de8e9facf57203879cf/packages/gatsby/src/bootstrap/load-plugins/validate.js#L97 and then we could skip whole ``if (isBad && process.env.NODE_ENV === `production`)`` in `packages/gatsby/src/bootstrap/load-plugins/index.js`

**Comment by andrewnester** [comment]
Thanks for the comments! @DSchau I've replaced other usage of `process.exit(1)`
@pieh addressed your comments as well

File: packages/gatsby/src/bootstrap/load-plugins/__tests__/validate.js [hunk 1]
```diff
@@ -93,30 +93,6 @@ describe(`handleBadExports`, () => {
...
-        ssr: [
-          {
-            exportName: `foo`,
-            pluginName: `default-site-plugin`,
-          },
-        ],
-      },
-    })
-    // TODO: snapshot console.log()'s from handleBadExports?
-    expect(result).toEqual(true)
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/bootstrap/load-plugins/__tests__/validate.js]: Check how I used mocking other `reporter` function here: https://github.com/gatsbyjs/gatsby/pull/8723/files#diff-f47b2b7f87ee42d8fcbe86b257a938ff

**Comment by andrewnester** [comment]
@pieh Thanks! I've updated code and added test back based on your comments.

File: packages/gatsby/src/bootstrap/load-plugins/__tests__/validate.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewnester) [reply packages/gatsby/src/bootstrap/load-plugins/__tests__/validate.js]: @pieh Thanks! Updated this test

**Comment by ChristopherBiscardi** [comment]
Hey @andrewnester @pieh, why was [this console.log](https://github.com/gatsbyjs/gatsby/pull/9031/files#diff-f61af2b51837961d211c7ea1a4b91837L294) changed to `panicOnBuild` in a patch release?

```diff
      if (!includesDefaultExport) {
-        console.log(``)
-        console.log(
+        report.panicOnBuild(
          `[${fileName}] The page component must export a React component for it to be valid`
        )
-        console.log(``)
      }
```

I see that there's a code comment saying "todo actually do die during builds", but changing to be a hard failure is breaking mdx's built-in page creator plugin support as of `v2.0.23`.

**Comment by pieh** [comment]
I didn't realize that mdx would be affected by this change and it seemed like valid change. We will revert that part

**Comment by KyleAMathews** [comment]
Or that check should be different for mdx? We'd want to die still for normal components.

**Comment by caulagi** [comment]
I had to downgrade gatsby due to this bug as well. I have the following reason component -

```
let component = ReasonReact.statelessComponent("404");

let styles =
  Css.{
    "container":
      style([
        alignContent(`center),
        display(`grid),
        height @@ vh(80.),
        justifyContent(`center),
        unsafe("justifyItems", "center"),
        textAlign(`center),
      ]),
  };

let make = _ => {
  ...component,
  render: _ =>
    <div className=styles##container>
      <Logo />
      <H1> <Text text="404!" /> </H1>
      <div> <Text text="Oops...that page doesn't exist!" /> </div>
    </div>,
};

let default = ReasonReact.wrapReasonForJs(~component, _ => make([||]));
```

It 'renders' to the following react component - 

```
import * as H1 from "../components/H1.bs.js";
import * as Css from "bs-css/src/Css.js";
import * as Logo from "../components/Logo.bs.js";
import * as Text from "../components/Text.bs.js";
import * as React from "react";
import * as ReasonReact from "reason-react/src/ReasonReact.js";

var component = ReasonReact.statelessComponent("404");

var styles = {
  container: Css.style(/* :: */[
        Css.alignContent(/* center */98248149),
        /* :: */[
          Css.display(/* grid */-999565626),
          /* :: */[
            Css.height(Css.vh(80)),
            /* :: */[
              Css.justifyContent(/* center */98248149),
              /* :: */[
                Css.unsafe("justifyItems", "center"),
                /* :: */[
                  Css.textAlign(/* center */98248149),
                  /* [] */0
                ]
              ]
            ]
          ]
        ]
      ])
};

function make() {
  return /* record */[
          /* debugName */component[/* debugName */0],
          /* reactClassInternal */component[/* reactClassInternal */1],
          /* handedOffState */component[/* handedOffState */2],
          /* willReceiveProps */component[/* willReceiveProps */3],
          /* didMount */component[/* didMount */4],
          /* didUpdate */component[/* didUpdate */5],
          /* willUnmount */component[/* willUnmount */6],
          /* willUpdate */component[/* willUpdate */7],
          /* shouldUpdate */component[/* shouldUpdate */8],
          /* render */(function () {
              return React.createElement("div", {
                          className: styles.container
                        }, ReasonReact.element(undefined, undefined, Logo.make(/* array */[])), ReasonReact.element(undefined, undefined, H1.make(undefined, /* array */[ReasonReact.element(undefined, undefined, Text.make("404!", /* array */[]))])), React.createElement("div", undefined, ReasonReact.element(undefined, undefined, Text.make("Oops...that page doesn't exist!", /* array */[]))));
            }),
          /* initialState */component[/* initialState */10],
          /* retainedProps */component[/* retainedProps */11],
          /* reducer */component[/* reducer */12],
          /* jsElementWrapped */component[/* jsElementWrapped */13]
        ];
}

var $$default = ReasonReact.wrapReasonForJs(component, (function () {
        return make(/* array */[]);
      }));

export {
  component ,
  styles ,
  make ,
  $$default ,
  $$default as default,

}
/* component Not a pure module */
```

`gatsby build` fails with this error -

```
success source and transform nodes — 0.018 s
success building schema — 0.169 s
success createPages — 0.000 s
error [<redacted>/src/pages/http404.bs.js] The page component must export a React component for it to be valid
error Command failed with exit code 1.
```

The page looks fine in the browser, so it seems to me there are two bugs -

1. In trying to find there is a component on the page,
1. and stopping the build by throwing a panic, as in this PR.

**Pull request merged**